internal/dcache/rpc/server/handler.go

```diff
@@ -47,43 +57,894 @@ var _ service.ChunkService = &ChunkServiceHandler{}
 }
 
 func NewChunkServiceHandler() *ChunkServiceHandler {
+	// TODO:: integration: get rvID, rvName and cache dir path for different RVs for the node from cluster manager
+	// below will be call to cluster manager to get the information
+	rvIDMap := getRvIDMap()
+
 	return &ChunkServiceHandler{
-		locks: common.NewLockMap(),
+		locks:   common.NewLockMap(),
+		rvIDMap: rvIDMap,
+	}
+}
+
+// check if the given mv is valid
+func (rv *rvInfo) isMvValid(mvPath string) error {
+	if !common.DirectoryExists(mvPath) {
+		return fmt.Errorf("MV path %s does not exist", mvPath)
+	}
+
+	mvName := filepath.Base(mvPath)
+	_, err := rv.getMVInfo(mvName)
+	return err
+}
+
+func (rv *rvInfo) getComponentRVsForMV(mvName string) []string {
+	mvInfo, err := rv.getMVInfo(mvName)
+	if err != nil {
+		log.Warn("ChunkServiceHandler::getComponentRVsForMV: Failed to get MV %s info for %s [%v]", mvName, rv.rvName, err.Error())
+		return nil
+	}
+
+	return mvInfo.componentRVs
+}
+
+// getMVInfo returns the mvInfo for the given mvName
+func (rv *rvInfo) getMVInfo(mvName string) (*mvInfo, error) {
+	val, ok := rv.mvMap.Load(mvName)
+	mvInfo := val.(*mvInfo)
+	if !ok || mvInfo == nil {
+		common.Assert(false, "MV %s not found in RV %s", mvName, rv.rvName)
+		return nil, fmt.Errorf("MV %s is invalid for RV %s", mvName, rv.rvName)
+	}
+
+	common.Assert(mvName == mvInfo.mvName, "MV name mismatch in mv", mvName, mvInfo.mvName)
+
+	return mvInfo, nil
+}
+
+func (rv *rvInfo) addToMVMap(mvName string, val *mvInfo) {
+	rv.mvMap.Store(mvName, val)
+	rv.mvCount.Add(1)
+}
+
+func (rv *rvInfo) deleteFromMVMap(mvName string) {
+	rv.mvMap.Delete(mvName)
+	rv.mvCount.Add(-1)
+}
+
+// return the current sync state of the MV
+func (mv *mvInfo) getSyncStatus() bool {
+	mv.mu.Lock()
+	defer mv.mu.Unlock()
+
+	return mv.isSyncing
+}
+
+// update the sync state of the MV
+func (mv *mvInfo) updateSyncState(isSyncing bool, syncID string, sourceRV string) error {
+	mv.mu.Lock()
+	defer mv.mu.Unlock()
+
+	if isSyncing && mv.isSyncing {
+		// if the source RV is different from the current source RV, it means that the current source RV's node is offline
+		// and is replaced by a new RV. So, the sync process needs to be restarted with the new RV.
+		if mv.sourceRV != sourceRV {
+			mv.syncID = syncID
+			mv.sourceRV = sourceRV
+		} else {
+			return fmt.Errorf("MV is already in syncing state with sync id %s and source RV %s", mv.syncID, mv.sourceRV)
+		}
+	}
+
+	mv.isSyncing = isSyncing
+	mv.syncID = syncID
+	mv.sourceRV = sourceRV
+
+	return nil
+}
+
+// update the component RVs for the MV
+func (mv *mvInfo) updateComponentRVs(componentRVs []string) {
+	mv.mu.Lock()
+	defer mv.mu.Unlock()
+
+	mv.componentRVs = componentRVs
+}
+
+// block new chunk operations for this MV till the sync is in progress
+func (mv *mvInfo) blockChunkOps() {
+	for {
+		if mv.blockOpsFlag.Load() {
+			time.Sleep(100 * time.Microsecond) // TODO: check if this is optimal
+		} else {
+			break
+		}
+	}
+}
```

Comment:
```
// Block the calling thread if this MV is currently quiesced, by StartSync or EndSync.
func (mv *mvInfo) blockIOIfMVQuiesced() error {
    // Common path.
    if ! mv.quiesceIOs.Load() {
        return nil
    }

    // Wait till MV is quiesced.
    now := time.Now()
    maxWait := 30 * time.Second

    for {
        if mv.quiesceIOs.Load() {
            time.Sleep(1 * time.Millisecond)

            elapsed := time.Since(now)
            if elapsed > maxWait {
                msg := fmt.Sprintf("%s still quiesced after %s", mv.mvName, maxWait)
                common.Assert(false, msg)
                return fmt.Errorf(msg)
            }
        } else {
            break
        }
    }

    return nil
}
```

Reply:
Done, updated

Reply:
Done